File: proy/proy-java/src/estrategia/Lancero.java
package estrategia;

public class Lancero extends Unidad {
	
	public static final int DAÑO = 25;
	
	public Lancero(Ubicacion ubicacion) {
		super(150, ubicacion);
	}

	public boolean atacar(Unidad unidad) {
		if (this.isMuerta() || !this.isDistanciaOK(unidad.getUbicacion(), 1, 3))
			return false;
		unidad.recibirDaño(DAÑO);
		return true;
	}

}

File: proy/src/polimorfismo/estrategia/lancero.py
"""Los lanceros pueden atacar a una distancia de 1 a 3, sin condición. 
Infringen un daño de 25 puntos, y comienzan con 150 de salud."""

from unidad import Unidad


class Lancero(Unidad):

    def __init__(self, posicion):
        super().__init__(25, 150, posicion)

    def puede_atacar(self, oponente):
        return (
            not self is oponente
            and not self.esta_muerto()
            and not oponente.esta_muerto()
            and self.esta_dentro_del_alcance(oponente, 1, 3)
        )

    def atacar(self, oponente):
        if self.puede_atacar(oponente):
            self.infringir_danio(oponente)
